src/clients/pagination/iter.rs

```diff
@@ -1,14 +1,15 @@
 //! Synchronous implementation of automatic pagination requests.
 
-use crate::client::{ClientError, ClientResult};
-use crate::model::Page;
+use crate::{model::Page, ClientError, ClientResult};
 
 /// Alias for `Iterator<Item = T>`, since sync mode is enabled.
-pub trait Paginator<T>: Iterator<Item = T> {}
-impl<T, I: Iterator<Item = T>> Paginator<T> for I {}
+pub type Paginator<'a, T> = Box<dyn Iterator<Item = T> + 'a>;
 
 /// This is used to handle paginated requests automatically.
-pub fn paginate<T, Request>(req: Request, page_size: u32) -> impl Iterator<Item = ClientResult<T>>
+pub fn paginate<'a, T: 'a, Request: 'a>(
```

Comment:
When removing them:

```
error[E0310]: the parameter type `T` may not live long enough
  --> src/clients/pagination/iter.rs:23:5
   |
9  | pub fn paginate<T, Request>(
   |                 - help: consider adding an explicit lifetime bound...: `T: 'static`
...
23 |     Box::new(pages.flat_map(|result| ResultIter::new(result.map(|page| page.items.into_iter()))))
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `FlatMap<PageIterator<Request>, ResultIte
r<T, std::vec::IntoIter<T>>, [closure@src/clients/pagination/iter.rs:23:29: 23:96]>` will meet its required lifetime bounds
```

You can try declaring `T` and `Request` as `'static`, but you'll get this:

```
error[E0759]: `self` has lifetime `'a` but it needs to satisfy a `'static` lifetime requirement
   --> src/clients/base.rs:264:13
    |
258 |           &'a self,
    |           -------- this data with lifetime `'a`...
...
264 | /             move |limit, offset| {
265 | |                 self.artist_albums_manual(artist_id, album_type, market, Some(limit), Some(offset))
266 | |             },
    | |_____________^ ...is captured here...
    |
note: ...and is required to live as long as `'static` here
   --> src/clients/base.rs:263:9
    |
263 |         paginate(
    |         ^^^^^^^^
```

So `Request` borrows the client, which means it can't be `'static`. It makes sense; the `Paginator` can't live longer than the client itself because of this, so its lifetime has to be bounded. I could make `T` static, if anything. I've added a quick note to clarify the lifetimes in the `src/clients/pagination/mod.rs` file.

Reply:
Great, it makes sense.